Add documentation for Result (#413)
icerpc/icerpc-docs: content/slice/language-guide/result-types.md

---
title: Result types
description: Learn how to define and use results in Slice.
---

{% slice1 %}

> The built-in generic type Result<Success, Failure> is not available with Slice1.

Reviewer: Again, I think it's good to wrap this in quotes, since it isn't a single word like 'sequence' or 'dictionary'.
It's a programmatic expression, not a grammatical expression.
```suggestion
The built-in generic type 'Result<Success, Failure>' is not available with Slice1.
```

Author: Fixed as suggested

{% /slice1 %}

{% slice2 %}

## Result

> A result is a built-in generic type that holds either a Success value or a Failure value, where `Success` and `Failure`

Reviewer: We're about the concept of results, not _a_ specific constructed result.
```suggestion
Result is a built-in generic type that either holds a Success value or a Failure value, where `Success` and `Failure`
```

are type parameters.

> You construct a result inline, without giving it a name, usually to specify the return type of an operation. For

Reviewer: This is the phrase we use on the sequence and dictionary pages, and I think it sounds better as well.
```suggestion
You can construct a result inline, without giving it a name, usually to specify the return type of an operation. For
```

Author: Fixes as suggested.

example:

```slice
module VisitorCenter

interface Greeter {
    greet(name: string) -> Result<string, GreeterError>
}

/// The errors returned by the Greeter.
unchecked enum GreeterError {
    EmptyName
    NameTooLong(maxLength: int32)
    Away(until: WellKnownTypes::TimeStamp)
}
```

> You can also use a constructed result as the type of a field, as the element type in a sequence, as the type of an

Reviewer: `of` is correct here, `in` would be for the actual elements themselves, not the their types.
Also, all the redundant `as`s sound strange to me.
```suggestion
You can also use a constructed result as the type of a field, the element type of a sequence, the type of an
```

Author: Fixed

operation parameter, and so on, just like any other Slice type.

A result is conceptually equivalent to a compact enum with two enumerators (`Success` and `Failure`), each with a single

> field with the associated type. You can see `Result<string, GreeterError>` as a convenient way to reference:

Reviewer: ```suggestion
field of the associated type. You can see `Result<string, GreeterError>` as a convenient way to reference:
```

Author: Fixed

```slice
compact enum Result<string, GreeterError> { // not a valid Slice identifier
    Success(value: string)
    Failure(value: GreeterError)
}
```

## C# mapping

> A `Result<Success, Failure>` maps to the a `Result<TSuccess, TFailure>`, where [Result<TSuccess, TFailure>] is a C#

Reviewer: I don't think this makes sense
```suggestion
A `Result<Success, Failure>` maps to `Result<TSuccess, TFailure>`, where [Result<TSuccess, TFailure>] is a C#
```

Author: Fixed

> record class in namespace ZeroC.Slice.

Reviewer: ```suggestion
record class in the ZeroC.Slice namespace .
```

Author: Fixed
